src/coreclr/jit/ssabuilder.cpp

```diff
@@ -853,6 +853,117 @@ void SsaBuilder::RenameDef(GenTreeOp* asgNode, BasicBlock* block)
+{
+    if (callNode->HasRetBufForStructArg())
+    {
+        // First check for "this"
+        GenTree* retBufArg = nullptr;
+        unsigned index     = 0;
+
+#ifdef DEBUG
+        bool foundThis        = false;
+        bool foundFtnAddr     = false;
+        bool renamedRetBufArg = false;
+#endif
+
+        if (callNode->gtCallThisArg != nullptr)
+        {
+#ifdef DEBUG
+            foundThis = true;
+#endif
+            index++;
+        }
+
+        for (GenTreeCall::Use& use : callNode->Args())
+        {
+            retBufArg = use.GetNode();
+
+            if (retBufArg->IsArgPlaceHolderNode())
+            {
+                // For place holder arg, find the corresponding late arg.
+                retBufArg = callNode->fgArgInfo->GetArgNode(index);
+            }
+
+            if (retBufArg->OperIs(GT_ASG))
+            {
+                // For setup arg, there might be an assignment node
+                // Get the RHS which could be the return buffer
+                retBufArg = retBufArg->AsOp()->gtGetOp2();
+            }
+
+            if (retBufArg->OperIs(GT_ADDR))
+            {
+                break;
+            }
+#ifdef DEBUG
+            else if ((retBufArg->gtFlags & GTF_ICON_FTN_ADDR) != 0)
+            {
+                foundFtnAddr = true;
+            }
+#endif
+            index++;
+        }
+
+        if ((retBufArg != nullptr) && retBufArg->OperIs(GT_ADDR))
+        {
+#ifdef DEBUG
+            // Verify that this is indeed the right arg that represents hidden return buffer
+            if (index == 2)
+            {
+                // func_call(this, ftnAddr, retBuffer, ...)
+                assert(foundThis && (foundFtnAddr || callNode->gtCallType == CT_INDIRECT));
+            }
+            else if (index == 1)
+            {
+                // func_call(<this or ftnAddr>, retBuffer, ...)
+                assert(foundThis || foundFtnAddr || callNode->gtCallType == CT_INDIRECT);
+            }
+            else
+            {
+                // func_call(retBuffer, ...)
+                assert(index == 0);
+            }
+#endif
+            GenTree*             addrOp1         = retBufArg->AsOp()->gtOp1;
+            GenTreeLclVarCommon* hiddenBufStruct = nullptr;
+            if (addrOp1->OperIs(GT_LCL_FLD))
+            {
+                hiddenBufStruct = addrOp1->AsLclFld();
+            }
+            else
+            {
+                assert(addrOp1->OperIs(GT_LCL_VAR));
+                hiddenBufStruct = addrOp1->AsLclVar();
+            }
+
+            LclVarDsc* varDsc = m_pCompiler->lvaGetDesc(hiddenBufStruct);
+            if (varDsc->IsHiddenBufferStructArg())
+            {
+                unsigned ssaNum = varDsc->lvPerSsaData.AllocSsaNum(m_allocator, block);
+
+                hiddenBufStruct->SetSsaNum(ssaNum);
+
+                m_renameStack.Push(block, hiddenBufStruct->GetLclNum(), ssaNum);
+                m_pCompiler->GetOpAsgnVarDefSsaNums()->Set(hiddenBufStruct, ssaNum);
```

Comment:
This is not correct: the partially-defined local should get the "use" number (see similar code in `RenameDef`).

Reply:
Actually, the use and def are at the same location, so won't they get same `SsaNum` for this case? It is different than `RenameDef` case.

Reply:
Hmm, not sure what you mean, both in `RenameDef`'s case and in `RenameCall`'s case use and def occur at the `ASG`/`CALL` respectively.

The fact we have two numbers is just to help VN not invalidate the whole local when we write only one field, it has no other purpose. Losing the use VN means VN won't be able to retrieve the "previous" map.

Reply:
I assume you are referring to this code in `RenameDef`:

https://github.com/dotnet/runtime/blob/ee2473239594817ae4cec1a6a63716820c7d25f2/src/coreclr/jit/ssabuilder.cpp#L773-L777

When you say `get the "use" number`, you are referring to `lclNode->SetSsaNum(m_renameStack.Top(lclNum));`, yes?

But for `CALL`, the stack is empty. What should be the right way to get the "use number" at this point?

Reply:
> But for `CALL`, the stack is empty

That is unexpected, it means something else in SSA is not working right for the local in question.

Reply:
OK. I think I have a repro for that. Let me track it down.

Reply:
Hmm, are you checking `lvaInSsa(hiddenBufStruct->GetLclNum())`? That might be the problem.

Reply:
That is missing.

Reply:
I realized that there is still a scenario where a stack could be empty. If the return struct is not used, we won't [initialize the stack](https://github.com/dotnet/runtime/blob/ee2473239594817ae4cec1a6a63716820c7d25f2/src/coreclr/jit/ssabuilder.cpp#L1375) with them.